### ISSUE TITLE
Add a simple way to add more teardown steps

### ISSUE THREAD

**Issue by Hubro** [robotframework/robotframework#4098]
When a test declares a `[Teardown]`, it overrides the default `Test Teardown` supplied in the test suite settings.

My team has been using Robot Framework for end-to-end testing for the past couple of years, and one trend I've seen over and over is the mass duplication of the default test teardown because several test cases need extra teardown steps.

For example:

```robot
*** Settings ***
Test Teardown    Run Keywords    Delete some temporary stuff
...              AND             Reset some database
...              AND             Restart some mock server

*** Test cases ***
Test that something works
    [Teardown]    Run Keywords    Delete some temporary stuff
    ...           AND             Reset some database
    ...           AND             Restart some mock server
    ...           AND             Also restore a thing that was deleted

Test that something writes some logs
    [Teardown]    Run Keywords    Delete some temporary stuff
    ...           AND             Reset some database
    ...           AND             Restart some mock server
    ...           AND             Also empty a log folder
```

As a result of this, we've started using a "Default Teardown" keyword, which improves the situation a lot:

```robot
*** Settings ***
Test Teardown    Default Teardown

*** Keywords ***
Default Teardown
    Delete some temporary stuff
    Reset some database
    Restart some mock server

*** Test cases ***
Test that something works
    [Teardown]    Run Keywords    Default Teardown
    ...           AND             Also restore a thing that was deleted

    # Do some testing

Test that something writes some logs
    [Teardown]    Run Keywords    Default Teardown
    ...           AND             Also empty a log folder

    # Do some testing
```

Still, this solution seems too wordy for a thing that (at least in my experience) is an extremely common problem.

Is there currently a cleaner (less wordy) way to achieve what we're doing?

If not, I have a couple of thoughts on how this could be improved. I would appreciate your feedback.

## Add a special keyword that refers to the default test teardown

Add a special keyword, for example `Run Default Test Teardown`, that resolves to the value of the default test teardown defined in the test suite settings.

Example usage:

```robot
*** Settings ***
Test Teardown    Run Keywords    Delete some temporary stuff
...              AND             Reset some database
...              AND             Restart some mock server

*** Test cases ***
Test that something works
    [Teardown]    Run Keywords    Run Default Test Teardown
    ...           AND             Also restore a thing that was deleted

    # Do some testing

Test that something writes some logs
    [Teardown]    Run Keywords    Run Default Test Teardown
    ...           AND             Also empty a log folder

    # Do some testing
```

This is still pretty wordy though, but at least there's no longer any need to create a default teardown keyword.

## Add [Before/After Teardown] settings to test cases

Add a setting to test cases that allows them to add a keyword call before and/or after the test teardown. This effectively allows test cases to add teardown logic without overriding the default teardown logic.

Example usage:

```robot
*** Settings ***
Test Teardown    Run Keywords    Delete some temporary stuff
...              AND             Reset some database
...              AND             Restart some mock server

*** Test cases ***
Test that something works
    [After Teardown]    Also restore a thing that was deleted

    # Do some testing

Test that something writes some logs
    [Before Teardown]   Also empty a log folder

    # Do some testing
```

**Comment by sebastianciupinski** [comment]
why cannot you just implement few helper keywords to handle different teardown scenarios?

according to before/after teardown - you can use listeners for that or use ListenerLibrary which implements listeners as keywords

**Comment by pekkaklarck** [comment]
I understand the problem but don't think that it's severe enough to warrant adding new settings. As noted above, listeners and also model modifiers could be used to manipulate data to add more steps to teardowns. Alternatively you could use something like this:

```robotframework
*** Settings ***
Test Teardown           Teardown

*** Variables ***
@{TEARDOWN}
...    Delete some temporary stuff
...    Reset some database
...    Restart some mock server

*** Test Cases ***
Test that something works
    # ...
    [Teardown]    After Teardown    Also restore a thing that was deleted

Test that something writes some logs
    # ...
    [Teardown]    Before Teardown    Also empty a log folder

*** Keywords ***
Teardown
    Run Keywords    @{TEARDOWN}

After Teardown
    [Arguments]    @{keywords}
    Run Keywords    @{keywords}
    Teardown

Before Teardown
    [Arguments]    @{keywords}
    Teardown
    Run Keywords    @{keywords}
```
All teardown keywords above are generic so you can put them into a resource file.

**Comment by Hubro** [comment]
@pekkaklarck I didn't know you could define and expand keyword arguments like that, that should make things a little easier. Thanks!
